File: usaco/chapter1/src/namenum.java
/*
ID: iamyfy1
PROG: namenum
LANG: JAVA
 */

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.PrintWriter;

/**
 * Created by yfy on 2016/11/2.
 */
public class namenum {
  private static int[] map = new int[]{2, 2, 2, 3, 3, 3, 4, 4, 4, 5, 5, 5,
      6, 6, 6, 7, -1, 7, 7, 8, 8, 8, 9, 9, 9, -1};

  private static boolean find;

  public static void main(String[] args) throws Exception {
    BufferedReader br = new BufferedReader(new FileReader("namenum.in"));
    String num = br.readLine();
    int numLen = num.length();
    br = new BufferedReader(new FileReader("dict.txt"));
    String word;
    PrintWriter pw = new PrintWriter("namenum.out");
    while ((word = br.readLine()) != null)
      if (match(num, word)) pw.println(word);
    if (!find) pw.println("NONE");
    pw.close();
  }

  private static boolean match(String num, String word) {
    if (num.length() != word.length()) return false;
    for (int i = 0; i < num.length(); i++)
      if (map[word.charAt(i) - 'A'] != num.charAt(i) - '0') return false;
    find = true;
    return true;
  }
}

File: usaco/cpp/namenum.cpp
/*
LANG: C++
PROG: namenum
*/
#include <fstream>
#include <string>

char letter2num(char ch)
{
	if (ch >= 'A' && ch <= 'C') return '2';
	if (ch >= 'D' && ch <= 'F') return '3';
	if (ch >= 'G' && ch <= 'I') return '4';
	if (ch >= 'J' && ch <= 'L') return '5';
	if (ch >= 'M' && ch <= 'O') return '6';
	if (ch >= 'P' && ch <= 'S') return '7';
	if (ch >= 'T' && ch <= 'V') return '8';
	if (ch >= 'W' && ch <= 'Y') return '9';
	return ' ';
}

int main()
{
	std::string num, num2, word;
	int i;
	char ch;
	bool none;

	std::ifstream fin("namenum.in");
	std::ofstream fout("namenum.out");
	std::ifstream fdict("dict.txt");

	fin >> num;
	none = true;
	while (fdict >> word)
	{
		if (num.size() == word.size())
		{
			num2 = "";
			for (i = 0; i < word.size(); i++)
			{
				ch = letter2num(word[i]);
				num2 += ch;
			}
			if (num == num2)
			{
				fout << word << '\n';
				none = false;
			}
		}
	}
	if (none) fout << "NONE\n";

	return 0;
}